morf-core/src/main/java/org/alfasoftware/morf/upgrade/UpgradePathFinder.java

```diff
@@ -214,6 +214,119 @@ private boolean schemasNotMatch(Schema targetSchema, Schema trialSchema, String
+   * Additionally, it logs a warning for any upgrade step with a UUID mismatch in the audit map.
+   *
+   * @param upgradeAudit A Map representing the upgrade audit information, where keys are step names and
+   *                     values are corresponding UUIDs.
+   * @throws IllegalStateException If discrepancies are found between upgrade steps and their associated UUIDs in the
+   *                               provided map.
+   */
+  public void findDiscrepancies(Map<String, String> upgradeAudit) {
+
+    List<String> discrepancies = stepsToApply.stream()
+            .filter(step -> isStepDuplicated(step, upgradeAudit))
+            .map(step -> step.getUuid().toString())
+            .collect(Collectors.toList());
+
+    if (!discrepancies.isEmpty()) {
+      String message = String.format("Some upgrade steps could have run before with different UUID. The following have been detected:%n%s",
+              String.join("\n", discrepancies));
+      log.error(message);
+      throw new IllegalStateException(message);
+    }
+
+    stepsToApply.stream()
+            .filter(step -> upgradeAudit.containsValue(step.getUuid().toString()))
+            .filter(step -> checkNamesAreNotEqual(step, upgradeAudit))
+            .forEach(s -> log.warn(String.format("Upgrade step with uuid %s has a mismatched package/name!", s.getUuid())));
+
+  }
+
+  /**
+   * Checks if a given CandidateStep is duplicated based on its name or description in the upgradeAudit map.
+   * A CandidateStep is considered duplicated if it has the same name or description but a different UUID
+   * compared to the entries in the upgradeAudit map.
+   *
+   * @param step The CandidateStep to be checked for duplication.
+   * @param upgradeAudit A Map<String, String> representing the upgrade audit information, where keys are step names
+   *                     or descriptions, and values are corresponding UUIDs.
+   * @return {@code true} if the CandidateStep is duplicated, {@code false} otherwise.
+   */
+  private boolean isStepDuplicated(CandidateStep step, Map<String, String> upgradeAudit) {
+    String name = step.getName();
+    return upgradeAudit.containsKey(name) && !upgradeAudit.get(name).equals(step.getUuid().toString());
+  }
+
+  /**
+   * Checks whether the uuid associated with a given {@code CandidateStep} instance are not equal
+   * in the provided {@code upgradeAudit} map when the names appear in the upgradeAudit map.
+   *
+   * @param step The {@code CandidateStep} instance to check.
+   * @param upgradeAudit The map containing upgrade audit information (UUIDs and associated names).
+   * @return {@code true} if the names are not equal or if there are conflicts in the upgrade audit, {@code false} otherwise.
+   */
+  private boolean checkNamesAreNotEqual(CandidateStep step, Map<String, String> upgradeAudit) {
+    if (areAllValuesDistinct(upgradeAudit)) {

```

Comment:
You are recomputing this distinctness over and over again, which is somewhat expensive.

Reply:
Removed this code.